stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -561,13 +561,24 @@ public struct Set<Element : Hashable> :
   public func intersect<
     S : SequenceType where S.Generator.Element == Element
   >(sequence: S) -> Set<Element> {
-    let other = sequence as? Set<Element> ?? Set(sequence)
-    var newSet = Set<Element>()
-    for member in self {
-      if other.contains(member) {
-        newSet.insert(member)
+
+    // Attempt to iterate the smaller between `self` and `sequence`.
+    // If sequence is not a set, iterate over it because it may be single-pass.
+    var newSet: Set<Element>
+
+    if let other = sequence as? Set<Element> {
+      newSet = Set<Element>(minimumCapacity: min(other.count, count))
```

Comment:
I'm not sure if requesting capacity is the right thing here.  We don't have a guarantee that the result will have that many elements.  The result could even be empty.